components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java

```diff
@@ -0,0 +1,349 @@
+                        getRESTAPITokenCache().put(jwtTokenIdentifier, jwtValidationInfo);
+                    }
+                    //Validating scopes
+                    log.info("JWT token validation success. starting scope validation");
+                    return handleScopeValidation(message, signedJWTInfo, accessToken);
+                } else {
+                    log.error("Invalid JWT token :" + getMaskedToken(jwtHeader));
+                    return false;
+                }
+
+            } else {
+                log.error("Invalid JWT token :" + getMaskedToken(jwtHeader));
+                return false;
+            }
+        } catch (ParseException e) {
+            log.error("Not a JWT token. Failed to decode the token. Reason: " + e.getMessage());
+        } catch (MalformedURLException e) {
+            log.error("Malformed URL found in request path.Reason: " + e.getMessage());
+        }
+        return false;
+    }
+
+    /**
+     * Handle scope validation
+     *
+     * @param accessToken   JWT token
+     * @param signedJWTInfo : Signed token info
+     * @param message       : cxf Message
+     */
+    private boolean handleScopeValidation(Message message, SignedJWTInfo signedJWTInfo, String accessToken)
+            throws ParseException {
+
+        OAuthTokenInfo oauthTokenInfo = new OAuthTokenInfo();
+        oauthTokenInfo.setAccessToken(accessToken);
+        oauthTokenInfo.setEndUserName(signedJWTInfo.getJwtClaimsSet().getSubject());
+
+        if (signedJWTInfo.getJwtClaimsSet().getClaim(APIConstants.JwtTokenConstants.SCOPE) != null) {
+
+            String orgId = message.get(RestApiConstants.ORGANIZATION).toString();
+            String[] scopes = signedJWTInfo.getJwtClaimsSet().getStringClaim(JwtTokenConstants.SCOPE)
+                    .split(JwtTokenConstants.SCOPE_DELIMITER);
+            scopes = java.util.Arrays.stream(scopes).filter(s -> s.contains(orgId))
+                    .map(s -> s.replace(APIConstants.URN_CHOREO + orgId + ":", ""))
+                    .toArray(size -> new String[size]);
+            oauthTokenInfo.setScopes(scopes);
+
+            if (validateScopes(message, oauthTokenInfo)) {
+                //Add the user scopes list extracted from token to the cxf message
+                message.getExchange().put(RestApiConstants.USER_REST_API_SCOPES, oauthTokenInfo.getScopes());
+                //If scope validation successful then set tenant name and user name to current context
+                String tenantDomain = MultitenantUtils.getTenantDomain(oauthTokenInfo.getEndUserName());
+                int tenantId;
+                PrivilegedCarbonContext carbonContext = PrivilegedCarbonContext.getThreadLocalCarbonContext();
+                RealmService realmService = (RealmService) carbonContext.getOSGiService(RealmService.class, null);
+                try {
+                    String username = oauthTokenInfo.getEndUserName(); //
```

Comment:
Empty comment start. Should be removed or complete?
```suggestion
                    String username = oauthTokenInfo.getEndUserName(); 
```